lib/services/local/headlessauthn_watcher_test.go

```diff
@@ -18,283 +18,306 @@ package local_test
-			// multiple events in quick succession, which could mark the waiter
-			// as stale. In practice, this stale behavior protects against race
-			// conditions, but in these these tests we want to control when the
-			// watcher is marked as stale.
-			First: 25 * time.Millisecond,
-			Step:  75 * time.Millisecond,
-			Max:   100 * time.Millisecond,
-		})
-		require.NoError(t, err)
+		// Make an update. Make sure we are servicing the updates channel first.
+		readyForUpdate := make(chan struct{})
+		stubC := make(chan *types.HeadlessAuthentication, 1)
+		go func() {
+			<-readyForUpdate
+			stub, err := s.identity.CreateHeadlessAuthenticationStub(ctx, pubUUID)
+			assert.NoError(t, err)
+			stubC <- stub
+		}()
 
-		timeoutCtx, timeoutCancel := context.WithTimeout(ctx, 500*time.Millisecond)
-		defer timeoutCancel()
+		ticker := time.NewTicker(time.Second)
+		t.Cleanup(ticker.Stop)
 
-		// We don't know when the waiter will be initialized, so we send
-		// events on each tick until one is received by the waiter.
 		for {
 			select {
-			case <-timeoutCtx.Done():
-				t.Fatal("Watcher never received an event")
-			case <-retry.After():
-				retry.Inc()
-				s.buf.Emit(backend.Event{
-					Type: types.OpPut,
-					Item: *item,
-				})
-			case <-firstEventReceivedC:
+			case update := <-sub.Updates():
+				// We should receive the update.
+				require.Equal(t, <-stubC, update)
 				return
+			case <-sub.Stale():
+				t.Fatal("Expected subscriber to not be marked as stale")
+			case <-ticker.C:
+				t.Fatal("Expected subscriber to receive an update")
+			case readyForUpdate <- struct{}{}:
 			}
 		}
-	}
+	})
 
-	t.Run("WaitEventWithConditionMet", func(t *testing.T) {
+	t.Run("Stale", func(t *testing.T) {
 		t.Parallel()
-		s := newTestEnv(t)
+		s := newHeadlessAuthenticationWatcherTestEnv(t, clockwork.NewFakeClock())
 
-		headlessAuthnCh, firstEventReceivedC, errC := waitInGoroutine(ctx, t, s.watcher, pubUUID, func(ha *types.HeadlessAuthentication) (bool, error) {
-			return ha.User != "", nil
-		})
+		sub, err := s.watcher.Subscribe(ctx, pubUUID)
+		require.NoError(t, err)
+		t.Cleanup(sub.Close)
 
-		// Emit put event that passes the condition.
-		stub, err := types.NewHeadlessAuthenticationStub(pubUUID, time.Now())
+		// Make an 2 updates without servicing the subscriber's Updates channel.
+		// The second update will be dropped and result in the subscriber being
+		// marked as stale.
+		stub, err := s.identity.CreateHeadlessAuthenticationStub(ctx, pubUUID)
+		require.NoError(t, err)
+		replace := *stub
+		replace.User = "user"
+		replace.PublicKey = []byte(sshPubKey)
+		_, err = s.identity.CompareAndSwapHeadlessAuthentication(ctx, stub, &replace)
 		require.NoError(t, err)
-		stub.User = "user"
 
-		waitForPutEvent(t, s, stub, firstEventReceivedC)
+		ticker := time.NewTicker(time.Second)
+		t.Cleanup(ticker.Stop)
 
-		require.NoError(t, <-errC)
-		require.Equal(t, stub, <-headlessAuthnCh)
+		select {
+		case <-sub.Stale():
+		case <-ticker.C:
```

Comment:
A ticker is best for use cases where you need it to tick multiple times. Looks like you're just in need of a single timeout here. Would time.After be a better tool for the job?

Reply:
I was trying to avoid a [memory leak issue with `time.After`](https://medium.com/@oboturov/golang-time-after-is-not-garbage-collected-4cbc94740082), though I'm not sure how important that is here. WDYT?

Reply:
It's not a leak, it's just that the timer won't be collected until after it expires. This matters if you'll be allocating many thousands of timers that won't expire until much later.

In your case:

1. The timer expires in a second.
2. This is a test, which means it's a short lived program. Even if it takes a second for the memory to be GC'd, your tests will likely have already completed by then and the process will have exited.